In [12]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import NearestNeighbors
from sklearn.decomposition import PCA
from sklearn.neighbors import NearestNeighbors
from sklearn.cluster import DBSCAN, KMeans

In [16]:
df_path = r"C:\Users\XT286AX\OneDrive - EY\Desktop\deposito__Montresori\datasets\train_mnist.csv"
df = pd.read_csv(df_path)
df.head()

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [17]:
X = df.drop("label", axis=1)
y = df[['label']]

scaler = StandardScaler()
x_scaled = scaler.fit_transform(X)

In [18]:
pca = PCA(n_components=0.95)
X_transformed = pca.fit_transform(x_scaled)
print(f"the number of components that allows a variance of 0.95 is {pca.n_components_} ")

the number of components that allows a variance of 0.95 is 320 


In [19]:
X_train, X_test, y_train, y_test = train_test_split(x_scaled, y, stratify=y, random_state=42)
X_train_pca, X_test_pca, y_train_pca, y_test_pca = train_test_split(X_transformed, y, stratify=y, random_state=42)


In [22]:

tree = DecisionTreeClassifier(random_state=42)

tree.fit(X_train_pca, y_train_pca)
y_pred_pca = tree.predict(X_test_pca)
print(f"the accuracy of the tree over the pca reduced data is: {accuracy_score(y_test, y_pred_pca)}")
print(f"the depth of the trained tree is: {tree.get_depth()}")

the accuracy of the tree over the pca reduced data is: 0.8064761904761905
the depth of the trained tree is: 41


In [23]:
tree.fit(X_train, y_train)
y_pred = tree.predict(X_test)
print(f"the accuracy of the tree over the pca reduced data is: {accuracy_score(y_test, y_pred)}")
print(f"the depth of the trained tree is: {tree.get_depth()}")

the accuracy of the tree over the pca reduced data is: 0.8563809523809524
the depth of the trained tree is: 42


Con la riduzione di dimensionalità da 784 features a 320, il decision tree classifier impiega più tempo (34 secondi vs 11 secondi) e ottiene un'accuracy minore. Nonostante la riduzione di dimensionalità porta ad una maggiore complessità per il classificatore, la profondità dell'albero praticamente non cambia (da una profondità di 41 con i dati pca a una di 42 con la totalità dei dati)